### DISCUSSION TITLE
matnwb: DynamicTable error in PlaneSegmentation object

### DISCUSSION THREAD

**Discussion by Icardij** [NeurodataWithoutBorders/helpdesk#28] [Q&A]
Hello,

I am very new to working with the NWB data format, and am currently having problems using matnwb. The long and short of it is when I run my script to create, and write the NWB file, it prints out the below error.

![Matlab code error](https://user-images.githubusercontent.com/93269545/165380423-c4a76c3a-6a6d-4211-bac1-d9e1d183743d.png)

I am not entirely sure on what a dynamic table is, though I do understand that the script seems to be crashing as it tries to create a PlaneSegmentation object that is later used to create a dynamic table.  (crash location posted below)

![Matlab code error](https://user-images.githubusercontent.com/93269545/165384123-f38a6b19-25f4-4712-bcbd-5150228b3790.png)

This code is a near exact copy of the Ophys tutorial base, except without the behavior section since we collect data from non-behaving subjects.
I went ahead and tried to use a exact copy of the Ophys tutorial (without behavior) and the same error appeared. If anyone has a idea of why this error is occurring, whether it's a incorrect argument in my script, a error in the install, or something else entirely I would appreciate the help.

Thanks

**Comment by oruebel** [comment]
> I am not entirely sure on what a dynamic table ...

NWB uses tables in a number of places to store metadata. DynamicTable is the base type that NWB uses to store tables. In a DynamicTable the data is stored in a column-based fashion, i.e., each column of the table is stored as a separate dataset/array. All DynamicTable objects have an ``id`` column and can otherwise contain an arbitrary number of columns and rows. The tables are called "Dynamic" because they allow addition of custom columns to include user-defined data without having to create an extension to the NWB data standard schema. 

The [PlaneSegmention](https://nwb-schema.readthedocs.io/en/latest/format.html#planesegmentation) type inherits from DynamicTable, i.e., PlaneSegmention is itself a table that defines a number of additional predefined columns.

> If anyone has a idea of why this error is occurring, ...

I'm not an expert in MatNWB, but based on the error message it appears that there is a mismatch between the id column and the rest of the columns in the PlaneSegmention table. Since PlaneSegmention is a table, all columns are required to have the same length. Based on the error, it seems that your ``id`` has length 1 while other columns are length 0. Is it possible that maybe your value for ``RG`` is wrong when you create the ``ElementIdentifiers`` for the table or maybe (since the table appears to be empty) you are missing some calls to add your data to the table?

CC @bendichter

**Reply by Icardij** [reply]
Correct me if I am wrong, but the PlaneSegmentation object should contain the image masks of each cell segmented from the data? If so than I would agree that the data is missing from the table, though I do not know how to add that in earlier. In the code isn't `plane_segmentation.image_mask = types.hdmf_common.VectorData( 'data', image_mask, 'description', 'image masks');` 
responsible for taking our masks, and adding it to the table? So should this come earlier than the `id` argument which is causing it to crash? And if so will I need to then add a command to create the `id` argument later? Or am I missing a earlier argument that needs to have our data in it?

**Reply by lawrence-mbf** [reply]
I believe you can set the `image_mask` property as a keyword argument and the validator should run correctly:
```MATLAB
plane_segmentation = types.core.PlaneSegmentation(...
  ...
  'id', types.hdmf_common.ElementIdentifiers(...),...
  'image_mask', types.hdmf_common.VectorData(...),...
  ...
);
```

**Reply by Icardij** [reply]
Alright, so I followed what you said and it worked for this object. Below is my current exact script, where "image_mask" is my matrix of pixel lists.
```plane_segmentation = types.core.PlaneSegmentation( ...  
'colnames', {'image_mask'}, ...  
'description', 'Cells segmented from visual cortex, Frames X cellnumber X pixellist', ...  
'image_mask', types.hdmf_common.VectorData('data', image_mask, 'description', 'image masks'), ...
'id', types.hdmf_common.ElementIdentifiers('data', int64(1:2000)), ...
'imaging_plane', types.untyped.SoftLink(imaging_plane));
```
   
Going on, it seems to kick back another error in the `roi_table_region` object, I'm attaching the error printout below. 

![Matlab code error](https://user-images.githubusercontent.com/93269545/165576648-b2d97c23-e3ca-4ebf-b365-833f10edc9d7.png)

I don't think this is related to the previous error, and instead think it is because of our actual data. `RG` which is what we have placed in the data property of the `roi_table_region` object is a ratio between red and green fluorescence instead of being a integer of just one fluorescence. I am assuming that is a problem for the table, though I may be very wrong. Should I open another discussion for this issue, or continue here?

**Reply by lawrence-mbf** [reply]
I think this is okay, we can keep the conversation here.

The roi_table_region object should be a `DynamicTableRegion` class whose data is actually the index into the table that it is referencing. I'm not sure what you would do if you're using it like a gradient other than setting each color used in the table.

CC @bendichter

**Reply by bendichter** [reply]
@Icardij could you copy/paste the entire command?

**Reply by Icardij** [reply]
Sure thing, There are a few variables that we define in our data, such as Im_both which is a 4D matrix of our raw images. (xpixel X ypixel X depth X channel) We also have image_mask, which is our segmented ROIs organized by (frame X cellnumber X pixellist) and is shared across both channels, and finally we have RG which is a ratio calculated between the values of the two channels. Not sure if any of these would cause problems. 

```matlab
nwb = NwbFile( ...
    'session_description', 'mouse brain imaging',...        
    'identifier', 'hSyn-AAV2/1-Addgene', ...                
    'session_start_time', datetime(2022, 1, 27), ...        
    'general_experimenter', 'Dan M', ...                       
    'general_institution', 'Lerner Research Institute', ...  
    'general_related_publications', 'TBD');                 

subject = types.core.Subject( ...
    'subject_id', 'D20', ...           
    'age', 'P90D', ...                 
    'description', 'AAV2/1 mouse', ...  
    'species', 'Mus musculus', ...      
    'sex', 'M' ...                      
);
nwb.general_subject = subject;

optical_channel = types.core.OpticalChannel( ...
    'description', 'two channels green/red', ...  
    'emission_lambda', 550);   

device = types.core.Device();
nwb.general_devices.set('Device', device); 

imaging_plane_name = 'imaging_plane';
imaging_plane = types.core.ImagingPlane( ...
    'optical_channel', optical_channel, ...  
    'description', 'Photoconversion of Neurons', ...  
    'device', types.untyped.SoftLink(device), ... 
    'excitation_lambda', 1100, ...  
    'imaging_rate', 0, ...  
    'indicator', 'CaMPARI', ...  
    'location', 'Visual cortex');  
nwb.general_optophysiology.set(imaging_plane_name, imaging_plane);

image_series = types.core.TwoPhotonSeries( ...
    'imaging_plane', types.untyped.SoftLink(imaging_plane), ... 
    'starting_time', 0.0, ... 
    'starting_time_rate', 3.0, ...
    'data', Im_both, ...                       
    'data_unit', 'unit16',...
    'description', 'Frames X Depth X Channel');
nwb.acquisition.set('TwoPhotonSeries', image_series);

plane_segmentation = types.core.PlaneSegmentation( ... 
    'colnames', {'image_mask'}, ...  
    'description', 'Cells segmented from visual cortex, Frames X cellnumber X pixellist', ...  
    'image_mask', types.hdmf_common.VectorData('data', image_mask, 'description', 'image masks'), ...
    'id', types.hdmf_common.ElementIdentifiers('data', int64(1:2000)), ...
    'imaging_plane', types.untyped.SoftLink(imaging_plane)); 
% plane_segmentation.image_mask = types.hdmf_common.VectorData( ...
%     'data', image_mask, 'description', 'image masks');                                          

img_seg = types.core.ImageSegmentation();
img_seg.planesegmentation.set('PlaneSegmentation', plane_segmentation);

ophys_module = types.core.ProcessingModule( ...
    'description',  'contains optical physiology data'); 
ophys_module.nwbdatainterface.set('ImageSegmentation', img_seg);
nwb.processing.set('ophys', ophys_module);

% Fluorescence and dF/F responces
roi_table_region = types.hdmf_common.DynamicTableRegion( ...
    'table', types.untyped.ObjectView(plane_segmentation), ...
    'description', 'Red to Green ratio from all ROIs', ...
    'data', RG);
roi_response_series = types.core.RoiResponseSeries( ...
    'rois', roi_table_region, ...
    'data', NaN(10, 100), ...
    'data_unit', 'Not Relevant', ...
    'description', 'Not Relevant',...
    'starting_time_rate', 3.0, ...
    'starting_time', 0.0);

fluorescence = types.core.Fluorescence();
fluorescence.roiresponseseries.set('RoiResponseSeries', roi_response_series);
ophys_module.nwbdatainterface.set('Fluorescence', fluorescence);

nwb.processing.set('ophys', ophys_module);

nwbExport(nwb, 'test.nwb');
```

**Reply by bendichter** [reply]
An "image_mask" should be a matrix that is n_rois * x * y. Could you reshape your data into that and try that? We also have the option of a "pixel mask" which is similar to what you are describing.

**Reply by Icardij** [reply]
I have changed the "image_mask" variable to what you said, as well as changing the 'data' argument in "roi_table_region" and "roi_response_series" so that the code now looks like below. 

```
roi_table_region = types.hdmf_common.DynamicTableRegion( ...
    'table', types.untyped.ObjectView(plane_segmentation), ...
    'description', 'all ROIs', ...
    'data', (1:length(RG))');
roi_response_series = types.core.RoiResponseSeries( ...
    'rois', roi_table_region, ...
    'data', RG', ...
    'data_unit', 'red to green ratio', ...
    'description', 'red to green ratio from same cell',...
    'starting_time_rate', 0.0, ...
    'starting_time', 0.0);
```
The script seems to run in its entirety now without kicking back any errors, and I wanted to check that this looks appropriate before I go ahead with actually upload our data.

**Reply by bendichter** [reply]
You should use 0-indexing for the `types.hdmf_common.DynamicTableRegion`: `'data', (0:length(RG)-1)'`
